src/test/java/uk/gov/justice/hmpps/prison/service/AppointmentsServiceImplTest.kt

```diff
@@ -0,0 +1,1241 @@
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.util.*
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.Function
+
+class AppointmentsServiceImplTest {
+  @Mock
+  private lateinit var bookingRepository: BookingRepository
+
+  @Mock
+  private lateinit var offenderBookingRepository: OffenderBookingRepository
+
+  @Mock
+  private lateinit var locationService: LocationService
+
+  @Mock
+  private lateinit var referenceDomainService: ReferenceDomainService
+
+  @Mock
+  private lateinit var scheduledAppointmentRepository: ScheduledAppointmentRepository
+
+  @Mock
+  private lateinit var offenderIndividualScheduleRepository: OffenderIndividualScheduleRepository
+
+  @Mock
+  private lateinit var agencyLocationRepository: AgencyLocationRepository
+
+  @Mock
+  private lateinit var agencyInternalLocationRepository: AgencyInternalLocationRepository
+
+  @Mock
+  private lateinit var eventStatusRepository: ReferenceCodeRepository<EventStatus>
+
+  @Mock
+  private lateinit var telemetryClient: TelemetryClient
+
+  private lateinit var appointmentsService: AppointmentsService
+
+  private val prison = AgencyLocation().apply { description = "description" }
+
+  @BeforeEach
+  fun initMocks() {
+    SecurityContextHolder.createEmptyContext()
+    ensureRoles(BULK_APPOINTMENTS_ROLE)
+    MockitoAnnotations.openMocks(this)
+
+    whenever(eventStatusRepository.findById(EventStatus.SCHEDULED_APPROVED)).thenReturn(
+      Optional.of(EventStatus("SCH", "desc")),
+    )
+    whenever(agencyLocationRepository.findById(any())).thenAnswer { invocation ->
+      Optional.of(prison.apply { id = invocation.getArgument<String>(0) })
+    }
+    whenever(offenderIndividualScheduleRepository.save(any())).thenAnswer { invocation ->
+      invocation.getArgument<OffenderIndividualSchedule>(0).apply { id = NEW_EVENT_ID }
+    }
+    whenever(offenderIndividualScheduleRepository.saveAndFlush(any())).thenAnswer { invocation ->
+      invocation.getArgument<OffenderIndividualSchedule>(0).apply { id = NEW_EVENT_ID }
+    }
+
+    appointmentsService = AppointmentsService(
+      offenderBookingRepository,
+      AuthenticationFacade(),
+      locationService,
+      referenceDomainService,
+      telemetryClient,
+      scheduledAppointmentRepository,
+      offenderIndividualScheduleRepository,
+      agencyLocationRepository,
+      agencyInternalLocationRepository,
+      eventStatusRepository,
+    )
+  }
+
+  @AfterEach
+  fun clearSecurityContext() {
+    SecurityContextHolder.clearContext()
+  }
+
+  @Nested
+  @DisplayName("Creating appointments")
+  internal inner class CreateAppointments {
+    @Test
+    fun createAppointments_returnsMultipleAppointmentDetails() {
+      stubLocation(LOCATION_B)
+      stubValidReferenceCode(REFERENCE_CODE_T)
+      stubValidBookingIds(LOCATION_B.agencyId, DETAILS_1.bookingId, DETAILS_2.bookingId)
+
+      val appointmentsToCreate = AppointmentsToCreate
+        .builder()
```

Comment:
I guess all the builders are because these are Java objects so can't be improved upon?

Reply:
Existing code 🤷 , I could have used a constructor since nearly all the fields are set i guess